Test cases for custom ops with autograd

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31282 Refactor custom op tests
* **#31003 Test cases for custom ops with autograd**

-

Differential Revision: [D18896189](https://our.internmc.facebook.com/intern/diff/D18896189/)

smessmer committed:
Test cases for custom ops with autograd

-

Differential Revision: [D18896189](https://our.internmc.facebook.com/intern/diff/D18896189/)

[ghstack-poisoned]

smessmer committed:
Update on "Test cases for custom ops with autograd"


-

Differential Revision: [D18896189](https://our.internmc.facebook.com/intern/diff/D18896189/)

[ghstack-poisoned]

smessmer committed:
Update on "Test cases for custom ops with autograd"


-

Differential Revision: [D18896189](https://our.internmc.facebook.com/intern/diff/D18896189/)

[ghstack-poisoned]

ezyang commented on test/custom_operator/test_custom_ops.cpp:
```diff
@@ -43,6 +43,51 @@ void get_operator_from_registry_and_execute() {
+void get_autograd_operator_from_registry_and_execute() {
+  auto& ops = torch::jit::getAllOperatorsFor(
+      torch::jit::Symbol::fromQualString("custom::op_with_autograd"));
+  AT_ASSERT(ops.size() == 1);
```
`TORCH_INTERNAL_ASSERT`

ezyang commented on test/custom_operator/test_custom_ops.cpp:
```diff
@@ -43,6 +43,51 @@ void get_operator_from_registry_and_execute() {
+  torch::jit::push(stack, x, 2, y);
+  op->getOperation()(stack);
+  torch::Tensor output = torch::jit::pop(stack).toTensor();
+  AT_ASSERT(0 == stack.size());
```
This is pretty boilerplatey, I think there's a simple utility function you could write to reduce the boilerplate quite a bit.

smessmer replied:
Added a new PR https://github.com/pytorch/pytorch/pull/31282 to also clean this up for pre-existing unit tests.

smessmer committed:
Update on "Test cases for custom ops with autograd"


-

Differential Revision: [D18896189](https://our.internmc.facebook.com/intern/diff/D18896189/)

[ghstack-poisoned]
